pkg/wasm/httpfetcher.go

```diff
@@ -0,0 +1,82 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+const (
+	defaultRetryBackoff = 50 * time.Millisecond
+)
+
+// HTTPFetcher fetches remote wasm module with HTTP get.
+type HTTPFetcher struct {
+	defaultClient *http.Client
+	retryBackoff  time.Duration
+}
+
+// NewHTTPFetcher create a new HTTP remote wasm module fetcher.
+func NewHTTPFetcher() *HTTPFetcher {
+	return &HTTPFetcher{
+		defaultClient: &http.Client{
+			Timeout: 5 * time.Second,
+		},
+		retryBackoff: defaultRetryBackoff,
+	}
+}
+
+// Fetch downloads a wasm module with HTTP get.
+func (f *HTTPFetcher) Fetch(url string, timeout *time.Duration) ([]byte, error) {
+	c := f.defaultClient
+	if timeout != nil {
+		c = &http.Client{
+			Timeout: *timeout,
+		}
+	}
+	attempts := 0
+	var lastError error
+	for attempts < 5 {
+		attempts++
+		resp, err := c.Get(url)
```

Comment:
This is a blocking call. If URL is 100MB, this might take a while. Can we do async download? Dismiss if you plan to have cache Get eventually be called from a goroutine.

Reply:
Yeah, this will be called from a goroutine.